Final Project Submission
Student Name: Benson Kamau
Student Pace: Full-Time
Instructor's: Nikita Njoroge

## Title: Predicting the functionality of Water wells in Tanzania

#### Overview

Tanzania, a country of nearly 57 million people, has a serious problem ensuring that its people have access to dependable and safe water sources. Although there are many water stations around the nation that have been created by the government and non-governmental organizations (NGOs), a significant number of these facilities are either non-operational or in critical need of repair. The objective of guaranteeing all Tanzanians long-term access to this essential resource is put in jeopardy by this circumstance. Ensuring the upkeep and efficiency of the current water infrastructure is of utmost importance in order to protect the welfare and advancement of people throughout this country in East Africa.

#### Challenges 

1. Environmental factors: Droughts and seasonal variations can have a big impact on well water supply. Extended dry spells can lower water levels, which can lead to wells being unreliable.
2. Contamination: A number of things, such as being next to a latrine, agricultural runoff, and naturally occurring contaminants in the groundwater, can pollute water wells. Communities that depend on these wells run the danger of health problems due to contaminated water.
3. Maintenance and repair: This can be attributed to factors including lack of funding for regular upkeep and lack of technical skills to maintain and repair the wells.
4. Public Awareness: The necessity of keeping water wells maintained and the potential health risks of drinking contaminated water are sometimes not well understood.

#### Proposed solution

1. Invest in boreholes and deeper wells that are more resilient to drought and seasonal fluctuations. Complementing these wells with other water sources would also help the community during dry season. 
2. Establish community-based water monitoring systems with the assistance of environmental and health organizations to guarantee routine supervision and timely reaction to pollution.
3. Encourage partnerships between the government, private sector, and NGOs to pool resources and share the responsibility of maintaining water wells.
4. Educate the community on the importance of maintaining the wells and risks associated with contaminated water.

#### Conclusion

The analysis identified key factors influencing well functionality, including geographic location, construction details, environmental conditions, and maintenance practices.

#### Problem Statement

Access to safe and consistent drinking water is a major challenge in Tanzania, especially in rural regions with minimal infrastructure. To solve this issue, the Tanzanian government has made investments in the building of water wells in collaboration with a number of non-governmental organizations. However, the sustainability and functionality of these wells remain uncertain, with many of them falling into disrepair or becoming non-functional over time.

#### Objectives

Main objective 
 1. To develop a machine learning classifier that predicts the condition of water wells in Tanzania.

Specific objectives
1. Prepare and analyse data for modeling
2. Develop predictive Model
3. Provide actionable insights and recommendations based on the findings of the model

### Importing relevant libraries 

In [366]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report,precision_score, recall_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

Define a class for model building

In [367]:
# Define my class
class Modelbuilder:
    def __init__(self, model):
        self.model = model

    def preprocessing(self, X_train, X_test):
        scaler = StandardScaler()
        X_train_scaled = scaler.fit_transform(X_train)
        X_test_scaled = scaler.transform(X_test)
        return X_train_scaled, X_test_scaled

    def fit(self, X_train, y_train):
        self.model.fit(X_train, y_train)    

    def predict(self, X_test):
        y_pred = self.model.predict(X_test)
        return y_pred
    
    def evaluate(self, y_test, y_pred):
        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred)
        recall = recall_score(y_test, y_pred)
        confusion_matrix = confusion_matrix(y_test, y_pred)
        return accuracy, precision, recall, confusion_matrix

### 1. Data Loading and Understanding

The data is sourced from Taarifa and the Tanzanian Ministry of Water. Data utilized can be found here: https://www.drivendata.org/competitions/7/pump-it-up-data-mining-the-water-table/data/

For the purposes of our evaluation, we are utilizing the Training Set Labels and Training Set Values, which include data from 59,400 pumps. 
We will then use the Test set values to test our models.

The following is a list of column names and descriptions:

* amount_tsh - Total static head (amount water available to waterpoint)
* date_recorded - The date the row was entered
* funder - Who funded the well
* gps_height - Altitude of the well
* installer - Organization that installed the well
* longitude - GPS coordinate
* latitude - GPS coordinate
* wpt_name - Name of the waterpoint if there is one
* num_private -
* basin - Geographic water basin
* subvillage - Geographic location
* region - Geographic location
* region_code - Geographic location (coded)
* district_code - Geographic location (coded)
* lga - Geographic location
* ward - Geographic location
* population - Population around the well
* public_meeting - True/False
* recorded_by - Group entering this row of data
* scheme_management - Who operates the waterpoint
* scheme_name - Who operates the waterpoint
* permit - If the waterpoint is permitted
* construction_year - Year the waterpoint was constructed
* extraction_type - The kind of extraction the waterpoint uses
* extraction_type_group - The kind of extraction the waterpoint uses
* extraction_type_class - The kind of extraction the waterpoint uses
* management - How the waterpoint is managed
* management_group - How the waterpoint is managed
* payment - What the water costs
* payment_type - What the water costs
* water_quality - The quality of the water
* quality_group - The quality of the water
* quantity - The quantity of water
* quantity_group - The quantity of water
* source - The source of the water
* source_type - The source of the water
* source_class - The source of the water
* waterpoint_type - The kind of waterpoint
* waterpoint_type_group - The kind of waterpoint

In [368]:
#create a function that loads data and gets the info about the data.
def load_and_get_info(file_path):
    """
    Load data from a CSV file and get information about the DataFrame.

    Parameters:
    - file_path (str): Path to the CSV file.

    Returns:
    - df_info (str): Information about the DataFrame.
    """
    # Load data
    df = pd.read_csv(file_path)

    # Display the first few rows of the DataFrame
    df_head = df.head()

    # Get information about the DataFrame
    df_info = df.info()

    return df,df_info, df_head

# A function that checks the data types of DataFrame columns and return the count of columns for each data type category.
def check_data_types(df):
    """
    Check the data types of DataFrame columns and return the count of columns for each data type category.

    Parameters:
    - df (DataFrame): Input DataFrame.

    Returns:
    - data_type_counts (dict): Dictionary containing the count of columns for each data type category.
    """
    data_type_counts = df.dtypes.replace({'object': 'string'}).value_counts().to_dict()
    return data_type_counts

#### 1.1 Loading our first dataset - Training set values

In [369]:
file_path1 = 'data/training set values.csv'
df1,data_info, data_head = load_and_get_info(file_path1)
print(data_info)
print("\nFirst few rows of the DataFrame:")
data_head #data_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55763 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59398 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    14/03/2011         Roman        1390         Roman   
1   8776         0.0    06/03/2013       Grumeti        1399       GRUMETI   
2  34310        25.0    25/02/2013  Lottery Club         686  World vision   
3  67743         0.0    28/01/2013        Unicef         263        UNICEF   
4  19728         0.0    13/07/2011   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... payment_type  \
0  34.938093  -9.856322                  none            0  ...     annually   
1  34.698766  -2.147466              Zahanati            0  ...    never pay   
2  37.460664  -3.821329           Kwa Mahundi            0  ...   per bucket   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...    never pay   
4  31.130847  -1.825359               Shuleni            0  ...    never pay   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   
3          soft          good           dry             dry   
4          soft          good      seasonal        seasonal   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   
3           machine dbh              borehole   groundwater   
4  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  
3  communal standpipe multiple    communal standpipe  
4           communal standpipe    communal standpipe  

[5 rows x 40 columns]

Our training set values contains 59400 rows and 40 columns. 

In [370]:
#check the data types of DataFrame columns in our training set values.
data_type_counts = check_data_types(df1)
print("Count of columns for each data type category:")
print(data_type_counts)

Count of columns for each data type category:
{'string': 30, dtype('int64'): 7, dtype('float64'): 3}


The dataset is divided into three data types categories:
1. String(object type) which has 30 columns e.g., funder, installer. 
2. Integer type which has 7 columns e.g., gpd_height, region_code, district_code
3. Float type which has 3 columns e.g., longitude and latitude

#### 1.2 Loading our second dataset - Training set labels

In [371]:
file_path2 = 'data/training set labels.csv'
df2,data_info, data_head = load_and_get_info(file_path2)
print(data_info)
print("\nFirst few rows of the DataFrame:")
data_head #data_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            59400 non-null  int64 
 1   status_group  59400 non-null  object
dtypes: int64(1), object(1)
memory usage: 928.3+ KB
None

First few rows of the DataFrame:


id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

The dataset has 2 columns(id and status_group) and 59,400 rows.

In [372]:
#check the data types of DataFrame columns in our training set values.
data_type_counts = check_data_types(df2)
print("Count of columns for each data type category:")
print(data_type_counts)


Count of columns for each data type category:
{dtype('int64'): 1, 'string': 1}


#### 1.3 Loading the third dataset - Test set values

In [373]:
file_path3 = 'data/Test set values.csv'
test_data,data_info, data_head = load_and_get_info(file_path3)
print(data_info)
print("\nFirst few rows of the DataFrame:")
data_head #data_head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14850 entries, 0 to 14849
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     14850 non-null  int64  
 1   amount_tsh             14850 non-null  float64
 2   date_recorded          14850 non-null  object 
 3   funder                 13980 non-null  object 
 4   gps_height             14850 non-null  int64  
 5   installer              13973 non-null  object 
 6   longitude              14850 non-null  float64
 7   latitude               14850 non-null  float64
 8   wpt_name               14850 non-null  object 
 9   num_private            14850 non-null  int64  
 10  basin                  14850 non-null  object 
 11  subvillage             14751 non-null  object 
 12  region                 14850 non-null  object 
 13  region_code            14850 non-null  int64  
 14  district_code          14850 non-null  int64  
 15  lg

id  amount_tsh date_recorded                  funder  gps_height  \
0  50785         0.0    04/02/2013                    Dmdd        1996   
1  51630         0.0    04/02/2013  Government Of Tanzania        1569   
2  17168         0.0    01/02/2013                     NaN        1567   
3  45559         0.0    22/01/2013              Finn Water         267   
4  49871       500.0    27/03/2013                  Bruder        1260   

    installer  longitude   latitude                 wpt_name  num_private  \
0        DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
1         DWE  36.656709  -3.309214                  Kimnyak            0   
2         NaN  34.767863  -5.004344           Puma Secondary            0   
3  FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
4      BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   

   ... payment_type water_quality quality_group      quantity  quantity_group  \
0  ...    never pay          soft          good      seasonal        seasonal   
1  ...    never pay          soft          good  insufficient    insufficient   
2  ...    never pay          soft          good  insufficient    insufficient   
3  ...      unknown          soft          good           dry             dry   
4  ...      monthly          soft          good        enough          enough   

                 source           source_type  source_class  \
0  rainwater harvesting  rainwater harvesting       surface   
1                spring                spring   groundwater   
2  rainwater harvesting  rainwater harvesting       surface   
3          shallow well          shallow well   groundwater   
4                spring                spring   groundwater   

      waterpoint_type waterpoint_type_group  
0               other                 other  
1  communal standpipe    communal standpipe  
2               other                 other  
3               other                 other  
4  communal standpipe    communal standpipe  

[5 rows x 40 columns]

The dataset has 14,850 rows and 40 columns.

In [374]:
#check the data types of DataFrame columns in our training set values.
data_type_counts = check_data_types(df3)
print("Count of columns for each data type category:")
print(data_type_counts)

Count of columns for each data type category:
{'string': 30, dtype('int64'): 7, dtype('float64'): 3}


The dataset is divided into three data types categories:
1. String(object type) which has 30 columns  
2. Integer type which has 7 columns 
3. Float type which has 3 columns 

#### 1.4 Merging the training set values and training set labels

In [375]:
# merging the training set values and training set labels on the 'id' column
train_data = pd.merge(df1, df2, on='id')
train_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55763 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59398 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

In [376]:
#Check the distribution of the target variable(status_group)
train_data['status_group'].value_counts()

status_group
functional                 32259
non functional             22824
functional needs repair     4317
Name: count, dtype: int64

### Data Cleaning

To start our data cleaning process, we will check the num_private column because no description has been provided.

In [377]:
train_data['num_private'].value_counts()

num_private
0       58643
6          81
1          73
5          46
8          46
        ...  
42          1
23          1
136         1
698         1
1402        1
Name: count, Length: 65, dtype: int64

Over 99% of the data = 0. I presume 0 means the data was not available and choose to drop the column entirely.

In [378]:
#drop the column 'num_private'
train_data.drop(columns=['num_private'],inplace=True)

I will assume that whoever recorded data did so truthfully and this should not effect our target. We drop the column.

In [379]:
#drop the column 'recorded_by'
train_data.drop(columns=['recorded_by'], inplace=True)

Next I explore columns with similar column names and descriptions to see if there is any overlap in ifnormation.

In [380]:
#Explore columns with similar column names and descriptions to see if there is any overlap in information.
def explore_similar_columns(df, col1, col2):
    """
    Explore columns with similar names in the DataFrame and return the value counts of the specified columns.

    Parameters:
    - df (DataFrame): The DataFrame containing the columns.
    - col1 (str): The name of the first column.
    - col2 (str): The name of the second column.

    Returns:
    - col1_value_counts (Series): Value counts of the first column.
    - col2_value_counts (Series): Value counts of the second column.
    """
    # Ensure the specified columns exist in the DataFrame
    if col1 not in df.columns or col2 not in df.columns:
        raise ValueError(f"One or both of the columns '{col1}' and '{col2}' do not exist in the DataFrame")
    
    # Get value counts for both columns
    col1_value_counts = df[col1].value_counts()
    col2_value_counts = df[col2].value_counts()

    return col1_value_counts, col2_value_counts

1. Payment and Payment_type

In [381]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'payment', 'payment_type')
print(col1_value_counts)

print(col2_value_counts)

payment
never pay                25348
pay per bucket            8985
pay monthly               8300
unknown                   8157
pay when scheme fails     3914
pay annually              3642
other                     1054
Name: count, dtype: int64
payment_type
never pay     25348
per bucket     8985
monthly        8300
unknown        8157
on failure     3914
annually       3642
other          1054
Name: count, dtype: int64


The two columns have similar information in the dataset. I will drop the payment type.

In [382]:
train_data.drop(columns=['payment_type'], inplace=True)

2. Quantity and quantity group

In [383]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'quantity', 'quantity_group')
print(col1_value_counts)

print(col2_value_counts)

quantity
enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: count, dtype: int64
quantity_group
enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: count, dtype: int64


The column 'quantity' and 'quantity_group' have similar names and descriptions, so I can drop one of them.

In [384]:
train_data.drop(columns=['quantity_group'], inplace=True)

3. Waterpoint_type and waterpoint_type group 

In [385]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'waterpoint_type', 'waterpoint_type_group')
print(col1_value_counts)

print(col2_value_counts)

waterpoint_type
communal standpipe             28522
hand pump                      17488
other                           6380
communal standpipe multiple     6103
improved spring                  784
cattle trough                    116
dam                                7
Name: count, dtype: int64
waterpoint_type_group
communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: count, dtype: int64


The above columns have similar names and descriptions, but one of the keys - communal standpipe has different values. I will keep the columns.

4. Source ,source_type and source_type_group

In [386]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'source', 'source_type')
print(col1_value_counts)

print(col2_value_counts)

source
spring                  17021
shallow well            16824
machine dbh             11075
river                    9612
rainwater harvesting     2295
hand dtw                  874
lake                      765
dam                       656
other                     212
unknown                    66
Name: count, dtype: int64
source_type
spring                  17021
shallow well            16824
borehole                11949
river/lake              10377
rainwater harvesting     2295
dam                       656
other                     278
Name: count, dtype: int64


In [387]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'source', 'source_class')
print(col1_value_counts)

print(col2_value_counts)

source
spring                  17021
shallow well            16824
machine dbh             11075
river                    9612
rainwater harvesting     2295
hand dtw                  874
lake                      765
dam                       656
other                     212
unknown                    66
Name: count, dtype: int64
source_class
groundwater    45794
surface        13328
unknown          278
Name: count, dtype: int64


I drop the source_type and source_class columns since they contain similar information to the source column, which is more robust.

In [388]:
train_data.drop(columns=['source_type', 'source_class'], inplace=True)

5. Extraction_type, extraction_typ_group,extraction_type_name

In [389]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'extraction_type', 'extraction_type_group')
print(col1_value_counts)

print(col2_value_counts)

extraction_type
gravity                      26780
nira/tanira                   8154
other                         6430
submersible                   4764
swn 80                        3670
mono                          2865
india mark ii                 2400
afridev                       1770
ksb                           1415
other - rope pump              451
other - swn 81                 229
windmill                       117
india mark iii                  98
cemo                            90
other - play pump               85
walimi                          48
climax                          32
other - mkulima/shinyanga        2
Name: count, dtype: int64
extraction_type_group
gravity            26780
nira/tanira         8154
other               6430
submersible         6179
swn 80              3670
mono                2865
india mark ii       2400
afridev             1770
rope pump            451
other handpump       364
other motorpump      122
wind-powered         117
india 

In [390]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'extraction_type', 'extraction_type_class')
print(col1_value_counts)

print(col2_value_counts)

extraction_type
gravity                      26780
nira/tanira                   8154
other                         6430
submersible                   4764
swn 80                        3670
mono                          2865
india mark ii                 2400
afridev                       1770
ksb                           1415
other - rope pump              451
other - swn 81                 229
windmill                       117
india mark iii                  98
cemo                            90
other - play pump               85
walimi                          48
climax                          32
other - mkulima/shinyanga        2
Name: count, dtype: int64
extraction_type_class
gravity         26780
handpump        16456
other            6430
submersible      6179
motorpump        2987
rope pump         451
wind-powered      117
Name: count, dtype: int64


I will keep the extraction_type and extraction_type_name columns. I will go ahead and drop the extraction_type_group column.

In [391]:
train_data.drop(columns=['extraction_type_group'], inplace=True)

6. Waterpoint_type and waterpoint_type_group

In [392]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'waterpoint_type', 'waterpoint_type_group')
print(col1_value_counts)

print(col2_value_counts)

waterpoint_type
communal standpipe             28522
hand pump                      17488
other                           6380
communal standpipe multiple     6103
improved spring                  784
cattle trough                    116
dam                                7
Name: count, dtype: int64
waterpoint_type_group
communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: count, dtype: int64


I will drop the water_type_group column

In [393]:
#drop the column 'waterpoint_type_group'
train_data.drop(columns=['waterpoint_type_group'], inplace=True)

7. Quality and Water_quality

In [394]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'quality_group', 'water_quality')
print(col1_value_counts)

print(col2_value_counts)

quality_group
good        50818
salty        5195
unknown      1876
milky         804
colored       490
fluoride      217
Name: count, dtype: int64
water_quality
soft                  50818
salty                  4856
unknown                1876
milky                   804
coloured                490
salty abandoned         339
fluoride                200
fluoride abandoned       17
Name: count, dtype: int64


I will drop the column 'quality_group' as it has the same values as the column 'water_quality'

In [395]:
#drop the column 'quality_group'
train_data.drop(columns=['quality_group'], inplace=True)

Next, we explore null values and decide how to clean nulls.

In [396]:
def check_nulls_and_duplicates(df):
    # Calculate the number of null values in each column
    null_counts = df.isnull().sum()

    # Calculate the total number of rows
    total_rows = len(df)

    # Calculate the percentage of null values in each column
    null_percentage = (null_counts / total_rows) * 100 
    
    # Display message about columns with null values
    columns_with_null = null_counts[null_counts > 0]
    if not columns_with_null.empty:
        print("Columns with null values and their count/percentage:")
        for column, count in columns_with_null.items():
            percentage = null_percentage[column]
            print(f"{column}: {count} ({percentage:.2f}%)")
    else:
        print("No null values")

    # Calculate the number of duplicate rows
    num_duplicates = df.duplicated().sum()

    # Display the number of duplicate rows
    print(f"Number of duplicate rows: {num_duplicates}")

    return num_duplicates

In [397]:
check_nulls_and_duplicates(train_data)

Columns with null values and their count/percentage:
funder: 3637 (6.12%)
installer: 3655 (6.15%)
wpt_name: 2 (0.00%)
subvillage: 371 (0.62%)
public_meeting: 3334 (5.61%)
scheme_management: 3878 (6.53%)
scheme_name: 28810 (48.50%)
permit: 3056 (5.14%)
Number of duplicate rows: 0


0

From the above output, i will explore the scheme_management and scheme_names further by checking the value counts of each.

In [398]:
col1_value_counts, col2_value_counts = explore_similar_columns(train_data, 'scheme_name', 'scheme_management')
print(col1_value_counts)

print(col2_value_counts)

scheme_name
K                       682
Borehole                546
Chalinze wate           405
M                       400
DANIDA                  379
                       ... 
Mradi wa maji Vijini      1
Villagers                 1
Magundi water supply      1
Saadani Chumv             1
Mtawanya                  1
Name: count, Length: 2695, dtype: int64
scheme_management
VWC                 36793
WUG                  5206
Water authority      3153
WUA                  2883
Water Board          2748
Parastatal           1680
Private operator     1063
Company              1061
Other                 766
SWC                    97
Trust                  72
Name: count, dtype: int64


I will drop scheme name since scheme management captures similar data more cleanly with fewer nulls. 

In [399]:
train_data.drop(columns=['scheme_name'], inplace=True)

I will go ahead and replace the null values in the coulmns with about 5% or 6% of null values with "unknown"

In [400]:
columns_to_fill = ['funder', 'installer', 'public_meeting', 'scheme_management']

for column in columns_to_fill:
    train_data[column] = train_data[column].fillna(value='Unknown')

I will also drop the null values in the wpt_name and subvillage columns.

In [401]:

train_data.dropna(subset=['wpt_name', 'subvillage'], inplace=True)

Finally, i will check the values in the permit column.

In [402]:
train_data['permit'].value_counts()

permit
True     38791
False    17180
Name: count, dtype: int64

I will replace the null values with True which is also the most common value

In [403]:
train_data['permit'] = train_data['permit'].fillna(value=True)

C:\Users\User\AppData\Local\Temp\ipykernel_36896\1024742864.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  train_data['permit'] = train_data['permit'].fillna(value=True)


In [404]:
check_nulls_and_duplicates(train_data)

No null values
Number of duplicate rows: 0


0

Next step is to check the year of construction

In [405]:
train_data['construction_year'].value_counts()

construction_year
0       20347
2010     2644
2008     2613
2009     2533
2000     2091
2007     1586
2006     1471
2003     1286
2011     1254
2004     1123
2012     1083
2002     1075
1978     1037
1995     1013
2005     1009
1999      978
1998      966
1990      954
1985      945
1980      811
1996      811
1984      779
1982      744
1994      738
1972      707
1974      676
1997      644
1992      640
1993      608
2001      539
1988      521
1983      488
1975      437
1986      434
1976      414
1970      411
1991      324
1989      316
1987      302
1981      238
1977      202
1979      192
1973      184
2013      176
1971      145
1960      102
1967       88
1963       85
1968       77
1969       59
1964       40
1962       30
1961       21
1965       19
1966       17
Name: count, dtype: int64

From the above output I can see that the construction year is 0 for 20347 rows. This is about 35% of our data.
I will find the median of the column without the 20347 rows.

In [406]:
# Create an object to hold column construction_year without year 0
train_data1 = train_data['construction_year'] 
train_data1.describe()

count    59027.000000
mean      1308.498297
std        949.092135
min          0.000000
25%          0.000000
50%       1986.000000
75%       2004.000000
max       2013.000000
Name: construction_year, dtype: float64

In [407]:
train_data1 = train_data1[train_data1 != 0]
train_data1.describe()

count    38680.000000
mean      1996.813056
std         12.472071
min       1960.000000
25%       1987.000000
50%       2000.000000
75%       2008.000000
max       2013.000000
Name: construction_year, dtype: float64

In [408]:
#Change the data type of construction_year to int
train_data['construction_year'] = train_data['construction_year'].astype(int)
# Replace 0 with 2000,
train_data['construction_year'] = train_data['construction_year'].replace(0, 2000)

In [409]:
#check the column construction_year after cleaning the data 
train_data['construction_year'].value_counts()

construction_year
2000    22438
2010     2644
2008     2613
2009     2533
2007     1586
2006     1471
2003     1286
2011     1254
2004     1123
2012     1083
2002     1075
1978     1037
1995     1013
2005     1009
1999      978
1998      966
1990      954
1985      945
1996      811
1980      811
1984      779
1982      744
1994      738
1972      707
1974      676
1997      644
1992      640
1993      608
2001      539
1988      521
1983      488
1975      437
1986      434
1976      414
1970      411
1991      324
1989      316
1987      302
1981      238
1977      202
1979      192
1973      184
2013      176
1971      145
1960      102
1967       88
1963       85
1968       77
1969       59
1964       40
1962       30
1961       21
1965       19
1966       17
Name: count, dtype: int64

I will then go ahead and create a column called age that has the age of the wells followd by dropping the contsruction year.


In [410]:
# create column age
train_data['age'] = 2024 - train_data['construction_year']

# drop column construction_year
train_data.drop(columns=['construction_year'], inplace=True)
train_data.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    14/03/2011         Roman        1390         Roman   
1   8776         0.0    06/03/2013       Grumeti        1399       GRUMETI   
2  34310        25.0    25/02/2013  Lottery Club         686  World vision   
3  67743         0.0    28/01/2013        Unicef         263        UNICEF   
4  19728         0.0    13/07/2011   Action In A           0       Artisan   

   longitude   latitude              wpt_name                    basin  ...  \
0  34.938093  -9.856322                  none               Lake Nyasa  ...   
1  34.698766  -2.147466              Zahanati            Lake Victoria  ...   
2  37.460664  -3.821329           Kwa Mahundi                  Pangani  ...   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu  Ruvuma / Southern Coast  ...   
4  31.130847  -1.825359               Shuleni            Lake Victoria  ...   

  extraction_type_class management  management_group         payment  \
0               gravity        vwc        user-group    pay annually   
1               gravity        wug        user-group       never pay   
2               gravity        vwc        user-group  pay per bucket   
3           submersible        vwc        user-group       never pay   
4               gravity      other             other       never pay   

  water_quality      quantity                source  \
0          soft        enough                spring   
1          soft  insufficient  rainwater harvesting   
2          soft        enough                   dam   
3          soft           dry           machine dbh   
4          soft      seasonal  rainwater harvesting   

               waterpoint_type    status_group  age  
0           communal standpipe      functional   25  
1           communal standpipe      functional   14  
2  communal standpipe multiple      functional   15  
3  communal standpipe multiple  non functional   38  
4           communal standpipe      functional   24  

[5 rows x 31 columns]

In [411]:
#We will check the region_code column
train_data['region_code'].value_counts()


region_code
11    5300
17    5010
12    4639
3     4378
5     4040
18    3324
19    3037
2     3024
16    2816
10    2640
4     2513
13    2093
14    1979
20    1969
1     1840
15    1808
6     1609
21    1583
80    1238
60    1025
90     917
7      805
99     423
9      390
24     326
8      300
40       1
Name: count, dtype: int64

In [412]:
# Drop the row with value as 40 since it will be hard to split the value
train_data = train_data[train_data['region_code'] != 40]

I will drop the following columns: iga, ward, region,id, management_group, source.

In [413]:
columns_to_drop = ['lga', 'ward', 'region', 'id', 'management_group', 'source', 'extraction_type', 'wpt_name', 'subvillage']

# Drop specified columns
train_data.drop(columns=columns_to_drop, inplace=True)

We will check the funders and installer columns in details.

In [414]:
#check the column funder, 50 most frequent values
train_data.funder.value_counts()[:50]

funder
Government Of Tanzania            9014
Unknown                           3641
Danida                            3114
Hesawa                            2200
Rwssp                             1374
Kkkt                              1286
World Vision                      1246
World Bank                        1244
Unicef                            1057
Tasaf                              876
District Council                   843
Dhv                                829
Private Individual                 825
Dwsp                               811
0                                  777
Norad                              765
Germany Republi                    610
Tcrs                               602
Ministry Of Water                  590
Water                              510
Dwe                                484
Netherlands                        470
Hifab                              450
Adb                                447
Lga                                442
Amref             

In [415]:
# Replace 0 with Unknown
train_data['funder'] = train_data['funder'].replace(to_replace='0', value='Unknown')

In [416]:
#check the column installer, 50 most frequent values
train_data.installer.value_counts()[:50]

installer
DWE                           17361
Unknown                        3658
Government                     1825
RWE                            1205
DANIDA                         1050
KKKT                            898
Commu                           894
Hesawa                          840
0                               777
TCRS                            707
Central government              622
CES                             610
Community                       553
DANID                           552
District Council                551
HESAWA                          539
LGA                             408
World vision                    408
WEDECO                          397
TASAF                           396
District council                392
Gover                           352
AMREF                           329
TWESA                           316
WU                              301
Dmdd                            287
ACRA                            278
World Vision      

In [417]:
# Replace 0 with Unknown
train_data['installer'] = train_data['installer'].replace(to_replace='0', value='Unknown')

In [418]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 59026 entries, 0 to 59399
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   amount_tsh             59026 non-null  float64
 1   date_recorded          59026 non-null  object 
 2   funder                 59026 non-null  object 
 3   gps_height             59026 non-null  int64  
 4   installer              59026 non-null  object 
 5   longitude              59026 non-null  float64
 6   latitude               59026 non-null  float64
 7   basin                  59026 non-null  object 
 8   region_code            59026 non-null  int64  
 9   district_code          59026 non-null  int64  
 10  population             59026 non-null  int64  
 11  public_meeting         59026 non-null  object 
 12  scheme_management      59026 non-null  object 
 13  permit                 59026 non-null  bool   
 14  extraction_type_class  59026 non-null  object 
 15  managem